### PR TITLE
Support rpm based targets plus data import  #7

### PR DESCRIPTION
**Pull request by artlog**
Support rocky / centos / redhat 
Add slapd-cli template
import initial data backup if provided ( ldaptoolbox_openldap_import_data_ldif )
Add syncrepl template to add on top of initial backup 
Add ppolicy template to add on top of initial backup 
Allow to provide localy certificates and keys to distribute

### CODE REVIEW

File: defaults/main.yml [hunk 1]
```diff
@@ -159,4 +161,16 @@ ldaptoolbox_openldap_overlay_refint_olcRefintNothing: "cn=nothing,{{ ldaptoolbox
 # Dynamic groups (dynlist)
 ldaptoolbox_openldap_overlay_dynlist_olcDlAttrSet: "groupOfURLs memberURL member+memberOf@groupOfNames*"
 
+#############
+# Data import
+#############
+
+ldaptoolbox_openldap_data_prefix: 'openldap-data'
+# if not defined the default population template will be used which is fine for test purposes
```
Comment (davidcoutadeur) [inline defaults/main.yml:169]: I have kept the import task, but IMO, it is not really a task which should be automatized

Comment (davidcoutadeur) [reply defaults/main.yml:169]: I have also added custom data to illustrate the behaviour

File: tasks/ldaptoolbox-add-ppolicy.yml [hunk 2]
```diff
@@ -0,0 +1,10 @@
+################################################################################
```
Comment (davidcoutadeur) [inline tasks/ldaptoolbox-add-ppolicy.yml:1]: For just one task, I have move this into main.yaml file

File: tasks/ldaptoolbox-add-syncrepl.yml [hunk 3]
```diff
@@ -0,0 +1,11 @@
+################################################################################
```
Comment (davidcoutadeur) [inline tasks/ldaptoolbox-add-syncrepl.yml:1]: For just one task, I have move this into main.yaml file

File: tasks/ldaptoolbox-repository.yml [hunk 4]
```diff
@@ -7,13 +7,39 @@
...
+        baseurl=https://ltb-project.org/rpm/openldap25/$releasever/$basearch
+        enabled=1
+        gpgcheck=1
+        gpgkey=file:///etc/pki/rpm-gpg/RPM-GPG-KEY-LTB-project
+      dest: /etc/yum.repos.d/ltb-project.repo
+  when: ( ((ansible_distribution == "CentOS")  or (ansible_distribution == "Rocky") or (ansible_distribution == 'RedHat' )) and ( ( ansible_distribution_major_version == "8" ) or ( ansible_distribution_major_version == "9" ) ) )
+
+- name: "unsupported distribution check"
+  ansible.builtin.shell: "echo '{{ ansible_distribution }} {{ ansible_distribution_major_version }} ' > /root/ansible_vars"
+  when: not ( (ansible_os_family == "Debian") or (ansible_distribution == "CentOS" and ansible_distribution_major_version == "8") )
```
Comment (davidcoutadeur) [inline tasks/ldaptoolbox-repository.yml:45]: I don't get the point here about `/root/ansible_vars`. In the rpm branch, I have replaced it with a warning message.
Also, I have added coherence with the previous check : the warning message should appear only if the system is unknown

File: tasks/main.yml [hunk 5]
```diff
@@ -56,6 +33,14 @@
     group: "{{ ldaptoolbox_openldap_configuration_group }}"
     mode: "{{ ldaptoolbox_openldap_configuration_mode }}"
 
+- name: deploy sladp-cli file
```
Comment (davidcoutadeur) [inline tasks/main.yml:36]: I have kept this in rpm branch, but it's a pity we have to maintain a whole new configuration file here.
It would have been simpler to change the few modification via a `lineinfile` task...

File: tasks/main.yml [hunk 6]
```diff
@@ -67,8 +52,30 @@
 - name: load config from file
   ansible.builtin.shell: "{{ ldaptoolbox_openldap_slapd_cli_cmd }} restoreconfig"
 
+- name: build generated data from template
+  ansible.builtin.shell: "{{ ldaptoolbox_openldap_slapd_cli_cmd }} importdatatemplate"
+  when: ldaptoolbox_openldap_generate_data_ldif is defined
```
Comment (davidcoutadeur) [inline tasks/main.yml:57]: I have kept this, but I really don't see the point of automating the import of sample data

File: tasks/main.yml [hunk 7]
```diff
@@ -67,8 +52,30 @@
...
+    src: "{{ ldaptoolbox_openldap_import_data_ldif }}"
+    dest: "{{ ldaptoolbox_openldap_configuration_backup_dir }}/{{ ldaptoolbox_openldap_data_prefix }}-{{ ldaptoolbox_openldap_configuration_timestamp }}.ldif"
+    owner: "{{ ldaptoolbox_openldap_configuration_owner }}"
+    group: "{{ ldaptoolbox_openldap_configuration_group }}"
+    mode: "{{ ldaptoolbox_openldap_configuration_mode }}"
+  when: ldaptoolbox_openldap_import_data_ldif is defined
+
+- name: add syncrepl account
+  include_tasks: ldaptoolbox-add-syncrepl.yml
+  when: ldaptoolbox_openldap_import_data_ldif is defined
```
Comment (davidcoutadeur) [inline tasks/main.yml:73]: This task is really specific. I have added a variable for not doing this every time, but only when needed.

Most of the customer will use either a complete file, or no file at all, waiting for the synchronization to occur.

File: tasks/main.yml [hunk 8]
```diff
@@ -67,8 +52,30 @@
...
+
+- name: add default ppolicy pwdPolicy
+  include_tasks: ldaptoolbox-add-ppolicy.yml
+  when: ldaptoolbox_openldap_import_data_ldif is defined
```
Comment (davidcoutadeur) [inline tasks/main.yml:77]: This task is really specific. I have added a variable for not doing this every time, but only when needed.

Most of the customer will use either a complete file, or no file at all, waiting for the synchronization to occur.

File: tasks/main.yml [hunk 9]
```diff
@@ -67,8 +52,30 @@
...
 
+- name: restore data
+  ansible.builtin.shell: "{{ ldaptoolbox_openldap_slapd_cli_cmd }} restore"
+  when: ( ldaptoolbox_openldap_import_data_ldif is defined ) or ( ldaptoolbox_openldap_generate_data_ldif is defined )
```
Comment (davidcoutadeur) [inline tasks/main.yml:81]: restoring data is useless and will overwrite the current data when doing a `ldaptoolbox_openldap_generate_data_ldif` task

File: templates/var/backups/openldap/config.ldif [hunk 10]
```diff
@@ -26,12 +27,16 @@ olcServerID: {{ ldaptoolbox_openldap_olcServerID }}
 olcSockbufMaxIncoming: 262143
 olcSockbufMaxIncomingAuth: 16777215
 olcThreads: 16
+{% if ldaptoolbox_openldap_olcTLSCACertificateFile %}
```
Comment (davidcoutadeur) [inline templates/var/backups/openldap/config.ldif:30]: This is modifing the current behaviour of `olcTLSProtocolMin`

I have fixed this in rpm branch.

File: templates/var/backups/openldap/config.ldif [hunk 11]
```diff
@@ -84,13 +89,15 @@ olcSortVals: {{ ldaptoolbox_openldap_olcSortVals }}
 dn: olcDatabase={0}config,cn=config
 objectClass: olcDatabaseConfig
 olcDatabase: {0}config
-olcAccess: {0}to *  by * none
+olcAccess: {0} to * by dn.exact=gidNumber=0+uidNumber=0,cn=peercred,cn=external,cn=auth manage by * none
 olcAddContentAcl: TRUE
 olcLastMod: TRUE
 olcMaxDerefDepth: 15
 olcReadOnly: FALSE
+{% if ldaptoolbox_openldap_config_olcRootDN %}
```
Comment (davidcoutadeur) [inline templates/var/backups/openldap/config.ldif:97]: This is OK, but IMO we must always define a rootDN for config backend when defined.

File: templates/var/backups/openldap/config.ldif [hunk 12]
```diff
@@ -100,6 +107,14 @@ objectClass: olcMdbConfig
 olcDatabase: {1}mdb
 olcDbDirectory: /usr/local/openldap/var/openldap-data
 olcSuffix: {{ ldaptoolbox_openldap_suffix }}
+{% if ldaptoolbox_openldap_database_olcAccess|length > 0 %}
```
Comment (davidcoutadeur) [inline templates/var/backups/openldap/config.ldif:110]: This was not needed because all the ACLs where defined in frontend backend.

This is ok because it adds a new way to do, but may be very confusing for admins understanding which rule is applied. (and which has precedence)

File: templates/var/backups/openldap/config.ldif [hunk 13]
```diff
@@ -111,12 +126,16 @@ olcRootPW: {{ ldaptoolbox_openldap_database_olcRootPW_hash }}
...
+olcSyncrepl: rid={{ syncrepl.rid }} provider={{ syncrepl.provider }} bindmethod=simple timeout=0 network-timeout=0 binddn="{{ syncrepl.binddn }}" credentials="{{ syncrepl.password }}" keepalive=0:0:0 starttls=no {% if syncrepl.tlscert %}tls_cert="{{ syncrepl.tlscert }}" tls_key="{{ syncrepl.tlskey }}" tls_cacert="{{ syncrepl.tlscacert }}" tls_reqcert="{{ syncrepl.tlsreqcert }}"{% endif %} filter="(objectclass=*)" searchbase="{{ syncrepl.searchbase }}" scope="{{ syncrepl.scope }}" schemachecking=on type="{{ syncrepl.type }}" retry="{{ syncrepl.retry }}"
 {% endfor %}
-{% if ldaptoolbox_openldap_syncrepl|length > 0 %}
+{% if ldaptoolbox_openldap_syncrepl|length > 1 %}
```
Comment (davidcoutadeur) [inline templates/var/backups/openldap/config.ldif:131]: Wrong.

A node can be multimaster even if configured with one replica.

File: templates/var/backups/openldap/ppolicy.ldif [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+pwdFailureCountInterval: 300
+pwdLockout: TRUE
+pwdLockoutDuration: 60
+pwdAllowUserChange: TRUE
+pwdExpireWarning: 864000
+pwdGraceAuthNLimit: 0
+pwdMustChange: TRUE
+pwdSafeModify: FALSE
+pwdCheckQuality: 1
+
```
Comment (davidcoutadeur) [inline templates/var/backups/openldap/ppolicy.ldif:26]: This is quite specific, but I kept it as an demonstration in standalone.yaml example.

File: templates/var/backups/openldap/syncrepl.ldif [hunk 15]
```diff
@@ -0,0 +1,20 @@
...
+uid: syncrepl
+objectClass: simpleSecurityObject
+objectClass: account
+{% if  ldaptoolbox_openldap_syncrepl_password_hash %}
+userPassword: {{ ldaptoolbox_openldap_syncrepl_password_hash }}
+{% else %}
+userPassword: {{ ldaptoolbox_openldap_syncrepl_password_vault }}
+{% endif %}
+# pwdPolicySubentry: cn=dsa,ou=ppolicies,{{ ldaptoolbox_openldap_suffix }}
+
```
Comment (davidcoutadeur) [inline templates/var/backups/openldap/syncrepl.ldif:20]: This is quite specific, but I kept it as an demonstration in standalone.yaml example.

**Comment by TonyBostonTB** [comment]
Any news on that one? Do you need help?

File: tasks/ldaptoolbox-repository.yml [hunk 16]
```diff
@@ -39,8 +38,7 @@
         gpgcheck=1
         gpgkey=file:///etc/pki/rpm-gpg/RPM-GPG-KEY-LTB-project
       dest: /etc/yum.repos.d/ltb-project.repo
-  when: (ansible_distribution == "CentOS" and ansible_distribution_major_version == "8")
-
+  when: ( ((ansible_distribution == "CentOS")  or (ansible_distribution == "Rocky") or (ansible_distribution == 'Red Hat Enterprise Linux' )) and ( ( ansible_distribution_major_version == "8" ) or ( ansible_distribution_major_version == "9" ) ) )
```
Comment (xavierba) [inline tasks/ldaptoolbox-repository.yml]: Rather than enumerating all (and possibly missing some, like Alma, Oracle, etc... here) flavors of RHEL like distros, one can use ``ansible_os_family == "RedHat"``

File: tasks/ldaptoolbox-repository.yml [hunk 17]
```diff
@@ -38,7 +38,7 @@
         gpgcheck=1
         gpgkey=file:///etc/pki/rpm-gpg/RPM-GPG-KEY-LTB-project
       dest: /etc/yum.repos.d/ltb-project.repo
-  when: ( ((ansible_distribution == "CentOS")  or (ansible_distribution == "Rocky") or (ansible_distribution == 'Red Hat Enterprise Linux' )) and ( ( ansible_distribution_major_version == "8" ) or ( ansible_distribution_major_version == "9" ) ) )
+  when: ( ((ansible_distribution == "CentOS")  or (ansible_distribution == "Rocky") or (ansible_distribution == 'RedHat' )) and ( ( ansible_distribution_major_version == "8" ) or ( ansible_distribution_major_version == "9" ) ) )
```
Comment (xavierba) [inline tasks/ldaptoolbox-repository.yml:41]: Use `ansible_os_family == "RedHat"` here too

File: README.md [hunk 18]
```diff
@@ -1,7 +1,7 @@
 OpenLDAP
 ========
 
-Ansible role which installs and configures [LTB-Project](https://ltb-project.org/)'s OpenLDAP on Debian target.
+Ansible role which installs and configures [LTB-Project](https://ltb-project.org/)'s OpenLDAP on Debian, CentOS, Rocky and RHEL targets.
```
Comment (xavierba) [inline README.md:4]: All Debian and RHEL clones or derivatives should be working, I'd rephrase as`Debian and RHEL like targets`

File: tasks/ldaptoolbox-repository.yml [hunk 19]
```diff
@@ -17,3 +17,31 @@
...
+        enabled=1
+        gpgcheck=1
+        gpgkey=file:///etc/pki/rpm-gpg/RPM-GPG-KEY-LTB-project
+      dest: /etc/yum.repos.d/ltb-project.repo
+  when: (ansible_distribution == "CentOS" and ansible_distribution_major_version == "8")
+
+
+- name: "unsupported distribution check"
+  ansible.builtin.shell: "echo '{{ ansible_distribution }} {{ ansible_distribution_major_version }} ' > /root/ansible_vars"
+  when: not ( (ansible_os_family == "Debian") or (ansible_distribution == "CentOS" and ansible_distribution_major_version == "8") )
```
Comment (xavierba) [inline tasks/ldaptoolbox-repository.yml:45]: `ansible_os_family == "RedHat"` here too

File: tasks/ldaptoolbox-repository.yml (hunk 17 above)
Comment (TonyBostonTB) [reply tasks/ldaptoolbox-repository.yml:41]: Maybe even go with `ansible_distribution_major_version >= "8"`

**Comment by artlog** [comment]
We will create a new branch rpm_only to get only RedHat like distribution support.

**Comment by davidcoutadeur** [comment]
A new branch (rpm_only) + a new pull request will be dedicated to this feature, as the current one (#8) mix multiple features.
